### DISCUSSION TITLE
Github Actions could benefit significantly from Dynamic Programming features

### DISCUSSION THREAD

**Discussion by damienwebdev** [community/community#38571] [Actions]
# Leverage Dynamic Programming w/ Github Actions 

It would be great if we could memoize our Github Actions and skip networking (and many other non-performant behaviors).

It would be nice if we could specify a "pureness" (interpreted to mean without side effects) for our actions so that if inputs to the action do not change, the action does not need to re-run and outputs from the task are just re-used.

This isn't a new topic, and is fairly well researched and understood (one could argue this is fundamental CS). I suspect, but obviously can't confirm, that this idea has been kicked around at Github but I don't know why it hasn't been implemented.

For those who are uninitiated on the topic, here's a lovely lecture from MIT specifically by @edemaine himself.

https://ocw.mit.edu/courses/6-006-introduction-to-algorithms-fall-2011/resources/lecture-14-depth-first-search-dfs-topological-sort/

## Caveats 

I understand that there are likely a lot of caveats, and not everyone will necessarily immediately understand this and could cause some confusion, but I think exposing `pure` on a step (or a more well-formed API) in combination with taking advantage of this at Github would be a massive performance win for all GA users. For historical purposes I would assume `pure` is unrequired by default and is also `false` by default.

```yml
steps:
  - uses: some-action@v3
    pure: true
```

**Comment by juliandunn** [comment]
I can't speak for the general case, but we are working on a performance optimization (internally called "skipping green trees") specifically around Git tree state -- that if your tree SHA is the same as what has been seen previously, we will just skip running the jobs and give you the same output & artifacts from the previous run. The use cases are a) on a merge to `main` you don't need to rebuild everything and b) if you revert a commit we can just give you the artifacts from the previous state.

**Reply by damienwebdev** [reply]
@juliandunn your idea is somewhat in the same vein, but my idea is perhaps a bit more fundamental, powerful, and broadly applicable to every commit in every codebase.

While I think the git tree feature is awesome and definitely a perf win, I would think that this generalization stands to address a separate need.

Consider, for example, a task called `longAction` which operates upon a large JSON document as an input and returns whether or not a certain object exists somewhere in that document.

More precisely, assume `longAction` is an action of 1 input, a JSON document, and 1 output, a boolean. Let us also assume that `longAction` is without side-effects, which can mean many things, but let's specifically consider that it doesn't operate with the network, doesn't operate on the filesystem, etc.

If the JSON document has not changed (and the action version has not changed), then output of this action must necessarily be the same.

This is a `pure` action. This kind of action (and I suspect that there are many such tasks in various actions repo) can essentially go from `Θ(n)` (or worse) to `Θ(1)`

However, purity is not true of many actions. Many wonderful existing actions operate on the filesystem, many operate upon the network, etc and this concept has limited application to those actions, thus the argument for `pure: false` by default.

Yet, even with that being the case, if we can turn a non-trivial percent of all actions into essentially dictionary lookups, we (and GitHub) can save hundreds of thousands of hours of compute (likely more).

There is an associated storage cost, ofc, because there's an associated cache, but I would argue it's a valuable feature.

**Reply by edemaine** [reply]
An important part of memoization is that you know what the inputs are, so you know whether they are the same as a previous run so the output can be re-used. I'm not a GH Actions expert; is this already specifiable somehow? If not, instead of a `pure: true` flag, perhaps you want something like `functionOf: [array of filenames or parameters]`.

I believe fancy JS build tools like TurboRepo and Nx already have features like this; they call it distributed caching. Notably, they have ways to define dependencies between steps so that those steps can be cached and re-used, like you're suggesting. It might be worth looking at their APIs to see how they could be adapted to GH Actions (probably better than my proposal above).

**Reply by spenserblack** [reply]
>is this already specifiable somehow?

I feel like `actions/cache` could meet a lot of memoization needs with some creative application. `actions/cache` with a key generated from `hashFiles('glob-pattern')` is pretty common, but probably could be expanded to "memoize" more general uses. For example, a cache key consisting of `${{ input.foo }}-${{ hashFiles('big-json-doc.json') }}-etc`, where you could use the cache on a cache hit, and perform the expensive operation if no cache is found. The limitation is that it would be *files* that are cached, but a step to parse that file into outputs could be used.

A downside is that downloading a cache can still be an expensive operation. I've seen \~0.5GB cache hang up a `windows-latest` job by several minutes. But you could trim down the cached items to help with this.

**Reply by damienwebdev** [reply]
@edemaine yes, inputs are already specifiable today:
 https://docs.github.com/en/actions/creating-actions/creating-a-composite-action#creating-an-action-metadata-file

The `pure` flag would be in addition to the already existing inputs to specify that a step has no side effects and that the memoization is "safe".

Interestingly, I already use NX for this purpose on several of my projects, but I would also love to see the concept implemented in GitHub Actions, TBH I would be a little surprised if it weren't already discussed.

@spenserblack if this were to occur in the action, then one would still need to download the action repo, etc. Many of these "little" things add up, which is why I think "first-party" infrastructure layer tooling may be a better fit for something this fundamental.
